### ISSUE TITLE
BrowserWindow not available in `will-download` item `done` event when quit during download

### ISSUE THREAD

**Issue by sindresorhus** [atom/electron#3488]
Run this and click one of the download links:

```js
'use strict';
const electron = require('electron');

electron.app.on('ready', () => {
	const win = new electron.BrowserWindow({});

	win.loadURL('https://github.com/sindresorhus/caprine/releases/tag/0.2.1');

	win.webContents.session.on('will-download', (e, item) => {
		item.on('done', () => {
			win.setProgressBar(-1);
		});
	});
});
```

Then close the window and you'll get this error:

<img width="532" alt="screen shot 2015-11-18 at 14 47 03" src="https://cloud.githubusercontent.com/assets/170270/11235439/3adf125a-8e04-11e5-8796-8850c89e1488.png">

It seems the `win` is no longer available when the `done` event is emitted with `cancelled` state.

I think the `done` event should be emitted before destroying the window.

**Comment by zcbenz** [comment]
The downloads are bound to session, so when the window or webContents is destroyed the downloading is still going. It is the same with downloading a file in Chrome, when you close the tab the downloading is still continued.

**Comment by sindresorhus** [comment]
Makes sense. Just didn't realize session could outlive the window.

So what's the best way to handle this? Currently, I'm listening to the window `close` event, but it's a bit verbose: https://github.com/sindresorhus/electron-dl/blob/506344214659e732ad7685ca499ced82085304bc/index.js#L21-L23

Maybe we could have a way to check if a window is still available to use. Like: `BrowserWindow#isOpen` or something?

**Comment by zcbenz** [comment]
I think in Chromium they just pauses all downloads when the app is going to quit, you can probably cancel all downloads when window is going to close.

There is a `isClosed` method can be used, though not documented.

**Comment by sindresorhus** [comment]
> There is a isClosed method can be used, though not documented.

Tried it now, but it inhibits the exact same problem and throws:

<img width="532" alt="screen shot 2015-11-18 at 14 47 03" src="https://cloud.githubusercontent.com/assets/170270/11235439/3adf125a-8e04-11e5-8796-8850c89e1488.png">

**Comment by zcbenz** [comment]
> Tried it now, but it inhibits the exact same problem and throws:

Oops, I have addressed it in #3510.
